File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081334Z_claude_claude-sonnet-5_php-symfony/src/Combat/CombatController.php
<?php

namespace App\Combat;

use App\Storage\KvStore;
use App\Support\Json;
use App\Support\Validators;
use Symfony\Component\HttpFoundation\JsonResponse;

/**
 * Handlers for initiative-tracking endpoints (/v1/combat/sessions/*).
 *
 * Sessions are stored as a single JSON document under the "combat_sessions"
 * kv_store key (see App\Storage\KvStore) rather than their own table, since
 * a session is small, always read/written as a whole, and never queried by
 * anything other than its id.
 */
final class CombatController
{
    private function withCombatState(callable $fn): mixed
    {
        return KvStore::withState('combat_sessions', ['sessions' => []], $fn);
    }

    private static function sessionActive(array $session): array
    {
        $entry = $session['order'][$session['turn_index']];
        return ['name' => $entry['name'], 'score' => $entry['score']];
    }

    private static function sessionSummary(array $session): array
    {
        return [
            'id' => $session['id'],
            'round' => $session['round'],
            'turn_index' => $session['turn_index'],
            'active' => self::sessionActive($session),
            'order' => array_map(static function ($entry) {
                return ['name' => $entry['name'], 'score' => $entry['score']];
            }, $session['order']),
        ];
    }

    public function createSession(array $body): JsonResponse
    {
        $id = $body['id'] ?? null;
        $combatants = $body['combatants'] ?? null;

        if (!is_string($id) || $id === '' || !is_array($combatants) || count($combatants) === 0) {
            return Json::error('invalid request');
        }

        $entries = [];
        foreach ($combatants as $combatant) {
            if (!is_array($combatant) || !isset($combatant['name'], $combatant['dex'], $combatant['roll'])
                || !is_string($combatant['name']) || !Validators::isValidInt($combatant['dex']) || !Validators::isValidInt($combatant['roll'])) {
                return Json::error('invalid request');
            }
            $dex = (int) $combatant['dex'];
            $roll = (int) $combatant['roll'];
            $entries[] = [
                'name' => $combatant['name'],
                'dex' => $dex,
                'score' => $roll + $dex,
            ];
        }

        usort($entries, static function ($a, $b) {
            if ($a['score'] !== $b['score']) {
                return $b['score'] <=> $a['score'];
            }
            if ($a['dex'] !== $b['dex']) {
                return $b['dex'] <=> $a['dex'];
            }
            return $a['name'] <=> $b['name'];
        });

        return $this->withCombatState(function (array &$state) use ($id, $entries) {
            if (isset($state['sessions'][$id])) {
                return Json::error('session already exists');
            }

            $conditions = [];
            foreach ($entries as $entry) {
                $conditions[$entry['name']] = [];
            }

            $session = [
                'id' => $id,
                'round' => 1,
                'turn_index' => 0,
                'order' => $entries,
                'conditions' => $conditions,
            ];

            $state['sessions'][$id] = $session;

            return new JsonResponse(self::sessionSummary($session));
        });
    }

    public function addCondition(array $body, string $id): JsonResponse
    {
        $target = $body['target'] ?? null;
        $condition = $body['condition'] ?? null;
        $duration = $body['duration_rounds'] ?? null;

        if (!is_string($target) || !is_string($condition) || !Validators::isValidInt($duration) || (int) $duration <= 0) {
            return Json::error('invalid request');
        }
        $duration = (int) $duration;

        return $this->withCombatState(function (array &$state) use ($id, $target, $condition, $duration) {
            if (!isset($state['sessions'][$id])) {
                return Json::error('session not found', 404);
            }
            $session = &$state['sessions'][$id];

            if (!isset($session['conditions'][$target])) {
                return Json::error('unknown target');
            }

            $session['conditions'][$target][] = [
                'condition' => $condition,
                'remaining_rounds' => $duration,
            ];

            return new JsonResponse([
                'target' => $target,
                'conditions' => $session['conditions'][$target],
            ]);
        });
    }

    public function advanceTurn(string $id): JsonResponse
    {
        return $this->withCombatState(function (array &$state) use ($id) {
            if (!isset($state['sessions'][$id])) {
                return Json::error('session not found', 404);
            }
            $session = &$state['sessions'][$id];

            $count = count($session['order']);
            $nextIndex = $session['turn_index'] + 1;
            if ($nextIndex >= $count) {
                $nextIndex = 0;
                $session['round'] += 1;
            }
            $session['turn_index'] = $nextIndex;

            $activeName = $session['order'][$nextIndex]['name'];
            $remaining = [];
            foreach ($session['conditions'][$activeName] as $cond) {
                $cond['remaining_rounds'] -= 1;
                if ($cond['remaining_rounds'] > 0) {
                    $remaining[] = $cond;
                }
            }
            $session['conditions'][$activeName] = $remaining;

            $conditionsOut = [];
            foreach ($session['conditions'] as $name => $conds) {
                $conditionsOut[$name] = $conds;
            }

            return new JsonResponse([
                'id' => $session['id'],
                'round' => $session['round'],
                'turn_index' => $session['turn_index'],
                'active' => self::sessionActive($session),
                'conditions' => empty($conditionsOut) ? new \stdClass() : $conditionsOut,
            ]);
        });
    }
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051117Z_claude_claude-sonnet-5_ruby-rails/app/controllers/combat_controller.rb
# Encounter combat sessions: initiative order, conditions, and turn
# advancement. Session state lives in COMBAT_SESSIONS (see
# lib/persistent_collections.rb) and is persisted after every mutation.
class CombatController < ApplicationController
  def create_combat_session
    id = params[:id]
    combatants = params[:combatants]

    unless id.is_a?(String) && !id.empty?
      render json: { error: 'invalid id' }, status: :bad_request
      return
    end

    if COMBAT_SESSIONS.key?(id)
      render json: { error: 'duplicate id' }, status: :bad_request
      return
    end

    unless combatants.is_a?(Array) && !combatants.empty?
      render json: { error: 'invalid combatants' }, status: :bad_request
      return
    end

    entries = []
    combatants.each do |c|
      name = c[:name]
      dex = c[:dex]
      roll = c[:roll]

      unless name.is_a?(String) && !name.empty? && valid_integer?(dex) && valid_integer?(roll)
        render json: { error: 'invalid combatant' }, status: :bad_request
        return
      end

      dex = dex.to_i
      roll = roll.to_i
      entries << { name: name, dex: dex, score: roll + dex }
    end

    entries.sort! { |a, b| compare_by_score_dex_name(a, b) }

    COMBAT_SESSIONS[id] = {
      order: entries,
      round: 1,
      turn_index: 0,
      conditions: {}
    }

    render json: combat_session_json(id)
  end

  def add_combat_condition
    session = COMBAT_SESSIONS[params[:id]]
    if session.nil?
      render json: { error: 'session not found' }, status: :not_found
      return
    end

    target = params[:target]
    condition = params[:condition]
    duration_rounds = params[:duration_rounds]

    unless target.is_a?(String) && session[:order].any? { |e| e[:name] == target }
      render json: { error: 'invalid target' }, status: :bad_request
      return
    end

    unless condition.is_a?(String) && !condition.empty?
      render json: { error: 'invalid condition' }, status: :bad_request
      return
    end

    unless valid_integer?(duration_rounds) && duration_rounds.to_i > 0
      render json: { error: 'invalid duration_rounds' }, status: :bad_request
      return
    end

    session[:conditions][target] ||= []
    session[:conditions][target] << { condition: condition, remaining_rounds: duration_rounds.to_i }
    COMBAT_SESSIONS.persist(params[:id])

    render json: {
      target: target,
      conditions: session[:conditions][target]
    }
  end

  def advance_combat_turn
    session = COMBAT_SESSIONS[params[:id]]
    if session.nil?
      render json: { error: 'session not found' }, status: :not_found
      return
    end

    order = session[:order]
    session[:turn_index] += 1
    if session[:turn_index] >= order.length
      session[:turn_index] = 0
      session[:round] += 1
    end

    active_name = order[session[:turn_index]][:name]
    active_conditions = session[:conditions][active_name]
    if active_conditions
      active_conditions.each { |c| c[:remaining_rounds] -= 1 }
      active_conditions.reject! { |c| c[:remaining_rounds] <= 0 }
    end

    conditions = session[:conditions]
    COMBAT_SESSIONS.persist(params[:id])

    active = order[session[:turn_index]]
    render json: {
      id: params[:id],
      round: session[:round],
      turn_index: session[:turn_index],
      active: { name: active[:name], score: active[:score] },
      conditions: conditions
    }
  end

  private

  def combat_session_json(id)
    session = COMBAT_SESSIONS[id]
    active = session[:order][session[:turn_index]]
    {
      id: id,
      round: session[:round],
      turn_index: session[:turn_index],
      active: { name: active[:name], score: active[:score] },
      order: session[:order].map { |e| { name: e[:name], score: e[:score] } }
    }
  end
end
